Symbol annotations on core

kohsuke opened the pull request:
(Continuing on from #2160)

To help [structs-plugin](https://wiki.jenkins-ci.org/display/JENKINS/Structs+plugin) and its users (such as pipeline, system-config-dsl, etc), we should have core classes annotated with `@Symbol`
### What changed from #2160?
- Targetting master branch, which is where the action is now.
- Additional symbol names are attached for f94ce3ac44ae47869daeceba76b01605de44c777
### What hasn't changed from #2160?
- Not pulling out changes that modify static field Descriptor instantiation like [this](https://github.com/jenkinsci/jenkins/pull/2160/files#r57040528)
- Not removing `@Symbol` on relatively minor extensions. I'd rather have one sweep to add this to wherever it can and avoid discovering few months down the road that the annotation was missing on this one class and it's 3 months away until the one line fix gets rolled into LTS
- `@Symbol` is on extensions, therefore it is on `Descriptor` and not on `Describable`. One instance where this got problematic is `ParameterValue` and we'll fix that by making `ParameterValue` nominally `Describable`, like we often do in similar cases (see [example](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/hudson/plugins/git/BranchSpec.java))
- `Jenkins.setViews()` is unrelated and not atomic, which is OK for the use case. I clarified that in javadoc.

@reviewbybees

kohsuke committed:
make the first view primary if none is chosen

... but some views already exist.

kohsuke committed:
Assigning some symbol names to use them

kohsuke committed:
More symbol assignments and constructor diet

to move more properties off to setters.

kohsuke committed:
More @Symbol annotations to test this out

kohsuke committed:
@Extension on static field is not kosher

kohsuke committed:
Adding @Symbol

Where applicable symbol names used by Job DSL plugin is used for maximum compatibility: https://jenkinsci.github.io/job-dsl-plugin/

kohsuke committed:
fixup field descriptor

kohsuke committed:
More symbols

kohsuke committed:
field descriptor fixup

kohsuke committed:
More symbol names

kohsuke committed:
More symbols

kohsuke committed:
descriptor fixup

kohsuke committed:
More symbols

kohsuke committed:
Field descriptor fixup

kohsuke committed:
More symbol annotations

kohsuke committed:
Rely on a released version

kohsuke committed:
Merge remote-tracking branch 'origin/2.0' into symbol

Conflicts:
	core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol.java
	core/src/main/java/jenkins/slaves/JnlpSlaveAgentProtocol2.java

kohsuke committed:
Changing this based on James Nord's suggestion

kohsuke committed:
This is not Describable, so no @DataBoundSetter here

kohsuke committed:
Renamed based on James Nord suggestion

kohsuke committed:
Merge remote-tracking branch 'origin/master' into symbol

Conflicts:
	core/src/main/java/hudson/model/FreeStyleProject.java
	core/src/main/java/jenkins/model/GlobalConfigurationCategory.java

kohsuke committed:
Added consistent names matching with other ParameterDescriptors

... based on comments in PR #2160

kohsuke committed:
Clarifying that this is not atomic and that's OK for the use case.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

kohsuke commented:
Calling out @daspilker as per requested.

Wrt `xxxParam`, I don't think we need to change it unless somebody brings that as a proposal. I have no problem with the current name, and as you say it solves the issue for `boolean`. I think reviewers have pointed it out just because the motivation wasn't obvious when looking at the diff.

kohsuke commented:
Turns out the test failure requires a small change in https://github.com/stapler/stapler/pull/74

daspilker commented:
+1 for adding the symbols.

The next Job DSL release will make use of symbols and the structs plugin. I tested this branch and it works well with Job DSL. I found some issues ([JENKINS-34105](https://issues.jenkins-ci.org/browse/JENKINS-34105), [JENKINS-34464](https://issues.jenkins-ci.org/browse/JENKINS-34464)), but nothing critical.

Is there some plan for further improvements in this area, e.g. sorting out mandatory vs optional parameters (aka `@DataBoundConstructor` vs `@DataBoundSetter`)? E.g. the `defaultValue` and `description` parameters are `@DataBoundConstructor` parameters for most `ParameterDefinition`s, so they are mandatory, but should not. Fixing this is IMHO an important second step for enabling concise scripts after adding the symbols.

KostyaSha commented:
👎  imho it should be enough to extend field in existing `@Extension` without adding new plugins into core distribution.

oleg-nenashev commented on core/src/main/java/hudson/cli/CliProtocol.java:
```diff
@@ -25,7 +26,7 @@
  * @since 1.467
  */
-@Extension
+@Extension @Symbol("cli")
```
How does it handle duplicated symbols?

kohsuke replied:
See [@Symbol javadoc](https://github.com/jenkinsci/structs-plugin/blob/master/annotation/src/main/java/org/jenkinsci/Symbol.java#L21-24) that clarifies this. (cc:@kzantow)

oleg-nenashev replied:
Uhh, there is some fun ahead when we start getting into conflicts

kzantow commented on core/src/main/java/hudson/FileSystemProvisioner.java:
```diff
@@ -235,7 +236,7 @@ public void restoreTo(AbstractBuild<?,?> owner, FilePath dst, TaskListener liste
         }
 
-        @Extension
+        @Extension @Symbol("default")
```
`default` is also a reserved keyword in Java.

kohsuke commented:
> Is there some plan for further improvements in this area, e.g. sorting out mandatory vs optional parameters (aka @DataBoundConstructor vs @DataBoundSetter)? E.g. the defaultValue and description parameters are @DataBoundConstructor parameters for most ParameterDefinitions, so they are mandatory, but should not. Fixing this is IMHO an important second step for enabling concise scripts after adding the symbols.

@daspilker, I haven't really thought about the further improvement beyond this initial round, but that one item you brought up makes a lot of sense. This PR has some changes in that area already that I have hit. I think you have some thoughts about what those further improvements should be. Can you send that as an email to the dev list so that we can discuss it there?

kohsuke commented:
@KostyaSha I considered that but rejected it. Making this a part of `@Extension` creates a real hindrance when we try to put symbols on plugins that define extensions, as it requires the plugin to rely on cutting-edge versions of the core. Note that this change does not add a new bundled plugin. It merely adds a new library jar file to the core.

kohsuke committed:
'boolean' is a not a valid Java identifier

kohsuke committed:
'default' is not a valid identifier, so we need something else.

This is truly unfortunate because 'default' is how we call them in the
UI. My choice is between something like 'default_' vs a synonym, so I
went for 'standard'

kohsuke commented:
Addressed the problem in the symbol names `default` and `boolean`.

kohsuke committed:
Merge remote-tracking branch 'origin/master' into symbol

rsandell commented:
:bee:

jglick commented:
:bee:
